llarp/router/outbound_message_handler.hpp

```diff
@@ -71,17 +89,41 @@ namespace llarp
     bool
     SendIfSession(const RouterID &remote, const Message &msg);
 
+    bool
+    QueueOutboundMessage(const RouterID &remote, Message &&msg,
+                         const PathID_t &pathid);
+
     void
-    FinalizeRequest(const RouterID &router, SendStatus status)
+    ProcessOutboundQueue();
+
+    void
+    RemoveEmptyPathQueues();
+
+    void
+    SendRoundRobin();
+
+    void
+    FinalizeSessionRequest(const RouterID &router, SendStatus status)
         LOCKS_EXCLUDED(_mutex);
 
-    mutable util::Mutex _mutex;  // protects outboundMessageQueue
+    llarp::thread::Queue< MessageQueueEntry > outboundQueue;
+    llarp::thread::Queue< PathID_t > removedPaths;
+    bool removedSomePaths;
+
+    mutable util::Mutex _mutex;  // protects pendingSessionMessageQueues
 
     std::unordered_map< RouterID, MessageQueue, RouterID::Hash >
-        outboundMessageQueue GUARDED_BY(_mutex);
+        pendingSessionMessageQueues GUARDED_BY(_mutex);
+
+    std::unordered_map< PathID_t, MessageQueue, PathID_t::Hash >
+        outboundMessageQueues GUARDED_BY(_mutex);
+
+    std::queue< PathID_t > roundRobinOrder;
 
     ILinkManager *_linkManager;
     std::shared_ptr< Logic > _logic;
+
+    const PathID_t zeroID;
```

Comment:
A comment explaining this magic value might help people who aren't aware of its purpose

Reply:
couldn't this be a static member?

Reply:
Will make static and add comment.

Reply:
Also, that reminds me that I need to make nodes reject paths with the zero ID, and path builds not use it.  Doing that now.